README.md
# screen-custom
Fork of GNU screen, with some personal QoL improvements


src/ChangeLog
Version 4.9.0 (30/01/2022):
  * Hardstatus option for used encoding (escape string '%e')
  * OpenBSD uses native openpty() from its utils.h
  * Fixes:
    - fix combining char handling that could lead to a segfault
    - CVE-2021-26937: possible denial of service via a crafted UTF-8 character sequence (bug #60030)
    - make screen exit code be 0 when checking --help
    - session names limit is 80 symbols (bug #61534)
    - option -X ignores specified user in multiuser env (bug #37437)
    - a lot of reformations/fixes/cleanups (man page and source code)

Version 4.8.0 (05/02/2020):
  * Improve startup time by only polling for files to close
  * Fixes:
	- Fix for segfault if termcap doesn't have Km entry
	- Make screen exit code be 0 when checking --version
	- Fix potential memory corruption when using OSC 49

Version 4.7.0 (02/10/2019):
  * Add support for SGR (1006) mouse mode
  * Add support for OSC 11
  * Update Unicode ambiguous and wide tables to 12.1.0
  * Fixes:
	- cross-compilation support (bug #43223)
	- a lot of manpage fixes and cleanups

Version 4.6.2 (23/10/2017):
  * Fixes:
	- revert changes to cursor position restore behavour (bug #51832)
	- set freed pointer to NULL (bug #52133)
	- documentation fixes
	- fix windowlist crashes (bug #43054 & #51500)

Version 4.6.1 (10/07/2017):
  * Fixes:
	- problems with starting session in some cases
	- parallel make install
	- segfault when querying info on nonUTF locale (bug #51402)

Version 4.6.0 (28/06/2017):
  * Update Unicode wide tables to 9.0 (bug #50044)
  * Support more serial speeds
  * Improved namespaces support
  * Migrate from fifos to sockets
  * Start viewing scrollback at first line of output (bug #49377)

Version 4.5.1 (25/02/2017):
  * Fixes:
	- logfile permissions problem (CVE-2017-5618)
	- SunOS build problem (bug #50089)
	- FreeBSD core dumps (bug #50143)

Version 4.5.0 (10/12/2016):
  * Allow specifying logfile's name via command line parameter '-L'
  * Fixes:
	- broken handling of "bind u digraph U+" (bug #48691)
	- crash with long $TERM (bug #48983)
	- crash when bumping blank window
	- build for AIX (bug #49149)
	- %x improperly separating arguments
	- install with custom DESTDIR (bug #48370)

Version 4.4.0 (19/06/2016):
  * Support up to 24 function keys
  * Fix runtime issues
  * 'logfile' command, starts logging into new file upon changing

Version 4.3.1 (28/06/2015):
  * Fix resize bug

Version 4.3.0 (13/06/2015):
  * Introduce Xx string escape showing the executed command of a window
  * Implement dead/zombie window polling, allowing for auto reconnecting
  * Allow setting hardstatus on first line
  * New Commands:
	- 'sort' command sorting windows by title
	- 'bumpleft', 'bumpright' - manually move windows on window list
	- 'collapse' removing numbering 'gaps' between windows, by renumbering
	- 'windows' command now accepts arguments for use with querying

Version 4.2.1 (28/04/2014):
  * allow for terminal with long $TERM (up to 32 characters)
  * allow to use long logins
  * documentation fixes
  * runtime fixes

Version 4.2.0 (17/04/2014):
  New Commands:
  * 'unbindall' to unbind all commands
  * 'up', 'down', 'left', 'right' sub-commands for 'focus'
  * 'rendition' to specify rendition to use in caption/hardstatus for window-names that have bell/monitor/silence/so turned on.
  * 'layout', with the following sub-commands
	- 'title'
	- 'number'
	- 'autosave' ('autosave on' or 'autosave off')
	- 'new'
	- 'save' ('save <name>')
	- 'select'
	- 'next'
	- 'prev'
	- 'attach'
	- 'show'
	- 'remove'
	- 'dump'
  * 'group' for moving window(s) into a group.
  * 'defmousetrack' and 'mousetrack', to turn on/off mouse-tracking for
    displays. It's turned off by default. With mouse-tracking turned on, it's
    possible to switch to a region ('focus') using mouse clicks. It's also
    possible to select a text region in copy-mode using a mouse click to place
    a mark and the scroll wheel to scroll through the buffer. Additional
    features might be to allow clicking on window-titles in the caption to
    switch to that window.
  * All commands prefixed '@' are treated as 'quiet', i.e. '@'-prefixed commands
    do not trigger any display messages.

  Changed Commands:
  * '-v' parameter to 'split' command for vertical splits.
  * 'sorendition' deprecated in favour of 'rendition so'.
  * 'digraph' can take a second parameter to specify custom digraphs. For
    example,
	'digraph >= ≥' or 'digraph >= U+2265'
    Using '0' as the second parameter will remove the digraph.
  * 'stuff' will prompt for input if there's nothing to stuff.
  * The argument to ":number" can be prefixed with '+' or '-' to use it as a
    relative argument.
  * '-g' parameter to 'windowlist' to show nested list of windows.
  * '//group' parameter to 'screen' to create a grouped window.
  * 'blankerprg' shows the currently set command on no argument.
  * 'maxwin' can now be used to increase the number of maximum windows.

  .screenrc:
  * $PID expands to the PID of the screen session.
  * $PWD expands to the current working directory of the session.
  * $STY expands to the session name.
  * Tilde-expansion in pathnames (e.g. for the 'source' command)
  * C-style escapes can be used (e.g. "\n" to get a newline with 'stuff')
  * '%p' in caption/hardstatus string expands to the PID of the backend, and
    '%+p' expands to the PID of the frontend (display).
  * '%S' in caption/hardstatus string expands to the session name.
  * '%P' in the caption string evaluates to true if the region is in copy mode.
  * '%E' in the caption string evaluates to true if the escape character has
    currently been pressed.

  Window List:
  * Nested views when there are window groups (with 'windowlist -g').
  * Press 'm' to toggle the most-recent view.
  * Press 'g' to toggle nestedness.
  * Press 'a' to view all windows in the list.
  * Press '/' to search in the list.
  * Press ',' and '.' to re-order windows in the list.

src/search.c
/* Copyright (c) 2008, 2009
 *      Juergen Weigert (jnweiger@immd4.informatik.uni-erlangen.de)
 *      Michael Schroeder (mlschroe@immd4.informatik.uni-erlangen.de)
 *      Micah Cowan (micah@cowan.name)
 *      Sadrul Habib Chowdhury (sadrul@users.sourceforge.net)
 * Copyright (c) 1993-2002, 2003, 2005, 2006, 2007
 *      Juergen Weigert (jnweiger@immd4.informatik.uni-erlangen.de)
 *      Michael Schroeder (mlschroe@immd4.informatik.uni-erlangen.de)
 * Copyright (c) 1987 Oliver Laumann
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (see the file COPYING); if not, see
 * https://www.gnu.org/licenses/, or contact Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02111-1301  USA
 *
 ****************************************************************
 */

#include <sys/types.h>

#include "config.h"
#include "screen.h"
#include "mark.h"
#include "extern.h"

#define INPUTLINE (flayer->l_height - 1)

extern struct layer *flayer;
extern struct win *fore;

#ifdef COPY_PASTE

int search_ic;

/********************************************************************
 *  VI style Search
 */

static int  matchword __P((char *, int, int, int));
static void searchend __P((char *, int, char *));
static void backsearchend __P((char *, int, char *));

void
Search(dir)
int dir;
{
  struct markdata *markdata;
  if (dir == 0)
    {
      markdata = (struct markdata *)flayer->l_data;
      if (markdata->isdir > 0)
	searchend(0, 0, NULL);
      else if (markdata->isdir < 0)
	backsearchend(0, 0, NULL);
      else
	LMsg(0, "No previous pattern");
    }
  else
    Input((dir > 0 ? "/" : "?"), sizeof(markdata->isstr)-1, INP_COOKED,
          (dir > 0 ? searchend : backsearchend), NULL, 0);
}

static void
searchend(buf, len, data)
char *buf;
int len;
char *data;	/* dummy */
{
  int x = 0, sx, ex, y;
  struct markdata *markdata;
  struct win *p;

  markdata = (struct markdata *)flayer->l_data;
  p = markdata->md_window;
  markdata->isdir = 1;
  if (len)
    strcpy(markdata->isstr, buf);
  sx = markdata->cx + 1;
  ex = flayer->l_width - 1;
  for (y = markdata->cy; y < p->w_histheight + flayer->l_height; y++, sx = 0)
    {
      if ((x = matchword(markdata->isstr, y, sx, ex)) >= 0)
        break;
    }
  if (y >= p->w_histheight + flayer->l_height)
    {
      LGotoPos(flayer, markdata->cx, W2D(markdata->cy));
      LMsg(0, "Pattern not found");
    }
  else
    revto(x, y);
}

static void
backsearchend(buf, len, data)
char *buf;
int len;
char *data;	/* dummy */
{
  int sx, ex, x = -1, y;
  struct markdata *markdata;

  markdata = (struct markdata *)flayer->l_data;
  markdata->isdir = -1;
  if (len)
    strcpy(markdata->isstr, buf);
  ex = markdata->cx - 1;
  for (y = markdata->cy; y >= 0; y--, ex = flayer->l_width - 1)
    {
      sx = 0;
      while ((sx = matchword(markdata->isstr, y, sx, ex)) >= 0)
	x = sx++;
      if (x >= 0)
	break;
    }
  if (y < 0)
    {
      LGotoPos(flayer, markdata->cx, W2D(markdata->cy));
      LMsg(0, "Pattern not found");
    }
  else
    revto(x, y);
}


/*
 * Search for a string that matches pattern.  The first character of the
 * match must be on line y, between columns sx and ex (inclusive), but
 * the rest of the match can extend beyond column ex and even onto the
 * following line.  Returns the starting column of the first match found,
 * or -1 if there's no match.
 */
static int
matchword(pattern, y, sx, ex)
char *pattern;
int y, sx, ex;
{
  unsigned char *cp, *pp, *cpe;
  int cy;

  fore = ((struct markdata *)flayer->l_data)->md_window;

  for (;sx <= ex; sx++)
    {
			cy = y;
			cp = WIN(cy)->image + sx;
			cpe = WIN(cy)->image + flayer->l_width;
      pp = (unsigned char *)pattern;
      for (;;)
	{
	  if (*cp != *pp)
	    if (!search_ic || ((*cp ^ *pp) & 0xdf) || (*cp | 0x20) < 'a' || (*cp | 0x20) > 'z')
	      break;
	  cp++;
	  pp++;
	  if (*pp == 0)
	    return sx;
		if (cp == cpe) {
		/*
		* We have a partial match, but we've hit
		* the end of this line.  Does it wrap onto
		* the following line?  If not, we're done.
		*/
		if (*cp == ' ' || cy >= fore->w_histheight + flayer->l_height - 1)
			break;

		/*
		* This line does wrap, so look on the next
		* line for the rest of our match.
		*/
		cy++;
		cp = WIN(cy)->image;
		cpe = WIN(cy)->image + flayer->l_width;
		}
	}
    }
  return -1;
}


/********************************************************************
 *  Emacs style ISearch
 */

static char *isprompts[] = {
  "I-search backward: ", "failing I-search backward: ",
  "I-search: ", "failing I-search: "
};


static int  is_redo __P((struct markdata *));
static void is_process __P((char *, int, char *));
static int  is_bm __P((char *, int, int, int, int));


static int
is_bm(str, l, p, end, dir)
char *str;
int l, p, end, dir;
{
  int tab[256];
  int i, q;
  unsigned char *s, c;
  int w = flayer->l_width;

  /* *sigh* to make WIN work */
  fore = ((struct markdata *)flayer->l_next->l_data)->md_window;
  debug2("is_bm: searching for %s len %d\n", str, l);
  debug3("start at %d end %d dir %d\n", p, end, dir);
  if (p < 0 || p + l > end)
    return -1;
  if (l == 0)
    return p;
  if (dir < 0)
    str += l - 1;
  for (i = 0; i < 256; i++)
    tab[i] = l * dir;
  for (i = 0; i < l - 1; i++, str += dir)
    {
      q = *(unsigned char *)str;
      tab[q] = (l - 1 - i) * dir;
      if (search_ic && (q | 0x20) >= 'a' && ((q | 0x20) <= 'z'))
        tab[q ^ 0x20] = (l - 1 - i) * dir;
    }
  if (dir > 0)
    p += l - 1;
  debug1("first char to match: %c\n", *str);
  while (p >= 0 && p < end)
    {
      q = p;
      s = (unsigned char *)str;
      for (i = 0;;)
	{
          c = (WIN(q / w))->image[q % w];
	  if (i == 0)
            p += tab[(int)(unsigned char) c];
	  if (c != *s)
	    if (!search_ic || ((c ^ *s) & 0xdf) || (c | 0x20) < 'a' || (c | 0x20) > 'z')
	      break;
	  q -= dir;
	  s -= dir;
	  if (++i == l)
	    return q + (dir > 0 ? 1 : -l);
	}
    }
  return -1;
}


/*ARGSUSED*/
static void
is_process(p, n, data)	/* i-search */
char *p;
int n;
char *data;	/* dummy */
{
  int pos, x, y, dir;
  struct markdata *markdata;

  if (n == 0)
    return;
  ASSERT(p);
  markdata = (struct markdata *)flayer->l_next->l_data;

  pos = markdata->cx + markdata->cy * flayer->l_width;
  LGotoPos(flayer, markdata->cx, W2D(markdata->cy));

  switch (*p)
    {
    case '\007':	/* CTRL-G */
      pos = markdata->isstartpos;
      /*FALLTHROUGH*/
    case '\033':	/* ESC */
      *p = 0;
      break;
    case '\013':	/* CTRL-K */
    case '\027':	/* CTRL-W */
      markdata->isistrl = 1;
      /*FALLTHROUGH*/
    case '\b':
    case '\177':
      if (markdata->isistrl == 0)
	return;
      markdata->isistrl--;
      pos = is_redo(markdata);
      *p = '\b';
      break;
    case '\023':	/* CTRL-S */
    case '\022': 	/* CTRL-R */
      if (markdata->isistrl >= (int)sizeof(markdata->isistr))
	return;
      dir = (*p == '\023') ? 1 : -1;
      pos += dir;
      if (markdata->isdir == dir && markdata->isistrl == 0)
	{
	  strcpy(markdata->isistr, markdata->isstr);
	  markdata->isistrl = markdata->isstrl = strlen(markdata->isstr);
	  break;
	}
      markdata->isdir = dir;
      markdata->isistr[markdata->isistrl++] = *p;
      break;
    default:
      if (*p < ' ' || markdata->isistrl >= (int)sizeof(markdata->isistr)
	  || markdata->isstrl >= (int)sizeof(markdata->isstr) - 1)
	return;
      markdata->isstr[markdata->isstrl++] = *p;
      markdata->isistr[markdata->isistrl++] = *p;
      markdata->isstr[markdata->isstrl] = 0;
      debug2("New char: %c - left %d\n", *p, (int)sizeof(markdata->isistr) - markdata->isistrl);
    }
  if (*p && *p != '\b')
    pos = is_bm(markdata->isstr, markdata->isstrl, pos, flayer->l_width * (markdata->md_window->w_histheight + flayer->l_height), markdata->isdir);
  if (pos >= 0)
    {
      x = pos % flayer->l_width;
      y = pos / flayer->l_width;
      LAY_CALL_UP
	(
          LayRedisplayLine(INPUTLINE, 0, flayer->l_width - 1, 0);
          revto(x, y);
          if (W2D(markdata->cy) == INPUTLINE)
	    revto_line(markdata->cx, markdata->cy, INPUTLINE > 0 ? INPUTLINE - 1 : 1);
        );
    }
  if (*p)
    inp_setprompt(isprompts[markdata->isdir + (pos < 0) + 1], markdata->isstrl ? markdata->isstr : "");
  flayer->l_x = markdata->cx;
  flayer->l_y = W2D(markdata->cy);
  LGotoPos(flayer, flayer->l_x, flayer->l_y);
  if (!*p)
    {
      /* we are about to finish, keep cursor position */
      flayer->l_next->l_x = markdata->cx;
      flayer->l_next->l_y = W2D(markdata->cy);
    }
}

static int
is_redo(markdata)
struct markdata *markdata;
{
  int i, pos, npos, dir;
  char c;

  npos = pos = markdata->isstartpos;
  dir = markdata->isstartdir;
  markdata->isstrl = 0;
  for (i = 0; i < markdata->isistrl; i++)
    {
      c = markdata->isistr[i];
      if (c == '\022')		/* ^R */
	pos += (dir = -1);
      else if (c == '\023')	/* ^S */
	pos += (dir = 1);
      else
	markdata->isstr[markdata->isstrl++] = c;
      if (pos >= 0)
	{
          npos = is_bm(markdata->isstr, markdata->isstrl, pos, flayer->l_width * (markdata->md_window->w_histheight + flayer->l_height), dir);
	  if (npos >= 0)
	    pos = npos;
	}
    }
  markdata->isstr[markdata->isstrl] = 0;
  markdata->isdir = dir;
  return npos;
}

void
ISearch(dir)
int dir;
{
  struct markdata *markdata;

  markdata = (struct markdata *)flayer->l_data;
  markdata->isdir = markdata->isstartdir = dir;
  markdata->isstartpos = markdata->cx + markdata->cy * flayer->l_width;
  markdata->isistrl = markdata->isstrl = 0;
  if (W2D(markdata->cy) == INPUTLINE)
    revto_line(markdata->cx, markdata->cy, INPUTLINE > 0 ? INPUTLINE - 1 : 1);
  Input(isprompts[dir + 1], sizeof(markdata->isstr) - 1, INP_RAW,
        is_process, NULL, 0);
  LGotoPos(flayer, markdata->cx, W2D(markdata->cy));
  flayer->l_x = markdata->cx;
  flayer->l_y = W2D(markdata->cy);
}

#endif /* COPY_PASTE */


src/window.h
/* Copyright (c) 2010
 *      Juergen Weigert (jnweiger@immd4.informatik.uni-erlangen.de)
 *      Sadrul Habib Chowdhury (sadrul@users.sourceforge.net)
 * Copyright (c) 2008, 2009
 *      Juergen Weigert (jnweiger@immd4.informatik.uni-erlangen.de)
 *      Michael Schroeder (mlschroe@immd4.informatik.uni-erlangen.de)
 *      Micah Cowan (micah@cowan.name)
 *      Sadrul Habib Chowdhury (sadrul@users.sourceforge.net)
 * Copyright (c) 1993-2002, 2003, 2005, 2006, 2007
 *      Juergen Weigert (jnweiger@immd4.informatik.uni-erlangen.de)
 *      Michael Schroeder (mlschroe@immd4.informatik.uni-erlangen.de)
 * Copyright (c) 1987 Oliver Laumann
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (see the file COPYING); if not, see
 * https://www.gnu.org/licenses/, or contact Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02111-1301  USA
 *
 ****************************************************************
 * $Id$ GNU
 */

#ifndef SCREEN_WINDOW_H
#define SCREEN_WINDOW_H

/* keep this in sync with the initialisations in window.c */
struct NewWindow
{
  int	StartAt;	/* where to start the search for the slot */
  char	*aka;		/* aka string */
  char	**args;		/* argv vector */
  char	*dir;		/* directory for chdir */
  char	*term;		/* TERM to be set instead of "screen" */
  int	aflag;
  int   dynamicaka;
  int	flowflag;
  int	lflag;
  int	histheight;
  int	monitor;
  int   wlock;		/* default writelock setting */
  int	silence;
  int   wrap;
  int	Lflag;		/* logging */
  int	slow;		/* inter character milliseconds */
  int   gr;
  int   c1;
  int   bce;
  int   encoding;
  char	*hstatus;
  char	*charset;
  int	poll_zombie_timeout;
};

#ifdef PSEUDOS

struct pseudowin
{
  int	p_fdpat;
  int	p_pid;
  int	p_ptyfd;
  struct event p_readev;
  struct event p_writeev;
  char	p_cmd[MAXSTR];
  char	p_tty[MAXSTR];
  char	p_inbuf[IOSIZE];	/* buffered writing to p_ptyfd */
  int	p_inlen;
};

/* bits for fdpat: */
#define F_PMASK 	0x0003
#define F_PSHIFT	2
#define F_PFRONT	0x0001			/* . */
#define F_PBACK 	0x0002			/* ! */
#define F_PBOTH 	(F_PFRONT | F_PBACK)	/* : */

#define F_UWP		0x1000			/* | */

/* The screen process ...)
 * ... wants to write to pseudo */
#define W_WP(w) ((w)->w_pwin && ((w)->w_pwin->p_fdpat & F_PFRONT))

/* ... wants to write to window: user writes to window 
 * or stdout/stderr of pseudo are duplicated to window */
#define W_WW(w) (!((w)->w_pwin) || \
(((w)->w_pwin->p_fdpat & F_PMASK) == F_PBACK) || \
((((w)->w_pwin->p_fdpat >> F_PSHIFT) & F_PMASK) == F_PBOTH) || \
((((w)->w_pwin->p_fdpat >> (F_PSHIFT * 2)) & F_PMASK) == F_PBOTH))

/* ... wants to read from pseudowin */
#define W_RP(w) ((w)->w_pwin && ((w)->w_pwin->p_fdpat & \
((F_PFRONT << (F_PSHIFT * 2)) | (F_PFRONT << F_PSHIFT)) ))

/* ... wants to read from window */
#define W_RW(w) (!((w)->w_pwin) || ((w)->w_pwin->p_fdpat & F_PFRONT))

/* user input is written to pseudo */
#define W_UWP(w) ((w)->w_pwin && ((w)->w_pwin->p_fdpat & F_UWP))

/* pseudo output has to be stuffed in window */
#define W_PTOW(w) (\
((w)->w_pwin->p_fdpat & F_PMASK << F_PSHIFT) == F_PBOTH << F_PSHIFT || \
((w)->w_pwin->p_fdpat & F_PMASK << F_PSHIFT * 2) == F_PBOTH << F_PSHIFT * 2 )

/* window output has to be stuffed in pseudo */
#define W_WTOP(w) (((w)->w_pwin->p_fdpat & F_PMASK) == F_PBOTH)

#endif /* PSEUDOS */

/* definitions for wlocktype */
#define WLOCK_OFF	0	/* all in w_userbits can write */
#define WLOCK_AUTO	1	/* who selects first, can write */
#define WLOCK_ON	2	/* user writes even if deselected */


#ifdef COPY_PASTE
struct paster
{
  char	*pa_pastebuf;		/* this gets pasted in the window */
  char	*pa_pasteptr;		/* pointer in pastebuf */
  int	 pa_pastelen;		/* bytes left to paste */
  struct layer *pa_pastelayer;	/* layer to paste into */
  struct event pa_slowev;	/* slowpaste event */
};
#else
struct paster;
#endif

struct win
{
  struct win *w_next;		/* next window */
  int    w_type;		/* type of window */
  void  *w_data;
  struct layer w_layer;		/* our layer */
  struct layer *w_savelayer;	/* the layer to keep */
  int    w_blocked;		/* block input */
#ifdef PSEUDOS
  struct pseudowin *w_pwin;	/* ptr to pseudo */
#endif
  struct display *w_pdisplay;	/* display for printer relay */
  struct display *w_lastdisp;	/* where the last input was made */
  int	 w_number;		/* window number */
  struct event w_readev;
  struct event w_writeev;
  struct event w_silenceev;	/* silence event */
  struct event w_zombieev;	/* event to try to resurrect window */
  int	 w_poll_zombie_timeout;
  int	 w_ptyfd;		/* fd of the master pty */
  char	 w_inbuf[IOSIZE];
  int	 w_inlen;
  char	 w_outbuf[IOSIZE];
  int	 w_outlen;
  int	 w_aflag;		/* (-a option) */
  int	 w_dynamicaka;		/* should we change name */
  char  *w_title;		/* name of the window */
  char  *w_akachange;		/* autoaka hack */
  char	 w_akabuf[MAXSTR];	/* aka buffer */
  int	 w_autoaka;		/* autoaka hack */
  struct win *w_group;		/* window group we belong to */
  int	 w_intermediate;	/* char used while parsing ESC-seq */
  int	 w_args[MAXARGS];	/* emulator args */
  int	 w_NumArgs;

#ifdef MULTIUSER
  int    w_wlock;		/* WLOCK_AUTO, WLOCK_OFF, WLOCK_ON */
  struct acluser *w_wlockuser;	/* NULL when unlocked or user who writes */
  AclBits w_userbits[ACL_BITS_PER_WIN];
  AclBits w_lio_notify;		/* whom to tell when lastio+seconds < time() */
  AclBits w_mon_notify;		/* whom to tell monitor statis */
#endif

  enum state_t w_state;		/* parser state */
  enum string_t w_StringType;
  struct mline *w_mlines;
  struct mchar w_rend;		/* current rendition */
#ifdef FONT
  char	 w_FontL;		/* character font GL */
  char	 w_FontR;		/* character font GR */
# ifdef ENCODINGS
  char	 w_FontE;		/* character font GR locked */
# endif
  int	 w_Charset;		/* charset number GL */
  int	 w_CharsetR;		/* charset number GR */
  int	 w_charsets[4];		/* Font = charsets[Charset] */
#endif
  int	 w_ss;
  struct cursor {
    int on;
    int	 x, y;
    struct mchar Rend;
#ifdef FONT
    int	 Charset;
    int	 CharsetR;
    int	 Charsets[4];
#endif
  } w_saved;
  int	 w_top, w_bot;		/* scrollregion */
  int	 w_wrap;		/* autowrap */
  int	 w_origin;		/* origin mode */
  int	 w_insert;		/* window is in insert mode */
  int	 w_keypad;		/* keypad mode */
  int	 w_cursorkeys;		/* appl. cursorkeys mode */
  int	 w_revvid;		/* reverse video */
  int	 w_curinv;		/* cursor invisible */
  int	 w_curvvis;		/* cursor very visible */
  int	 w_autolf;		/* automatic linefeed */
  char  *w_hstatus;		/* hardstatus line */
  int	 w_gr;			/* enable GR flag */
  int	 w_c1;			/* enable C1 flag */
  int	 w_bce;			/* enable backcol erase */
#if 0
  int    w_encoding;		/* for input and paste */
#endif
  int    w_decodestate;		/* state of our input decoder */
#ifdef DW_CHARS
  int    w_mbcs;		/* saved char for multibytes charset */
#endif
  char	 w_string[MAXSTR];
  char	*w_stringp;
  char	*w_tabs;		/* line with tabs */
  int	 w_bell;		/* bell status of this window */
  int	 w_flow;		/* flow flags */
  struct logfile *w_log;	/* log to file */
  int    w_logsilence;		/* silence in secs */
  int	 w_monitor;		/* monitor status */
  int	 w_silencewait;		/* wait for silencewait secs */
  int	 w_silence;		/* silence status (Lloyd Zusman) */
  char	 w_vbwait;
  char	 w_norefresh;		/* dont redisplay when switching to that win */
#ifdef RXVT_OSC
  char	 w_xtermosc[5][MAXSTR];	/* special xterm/rxvt escapes */
#endif
  int    w_mouse;		/* mouse mode 0,9,1000 */
  int    w_extmouse;		/* extended mouse mode 0,1006 */
#ifdef HAVE_BRAILLE
  int	 w_bd_x, w_bd_y;	/* Braille cursor position */
#endif

#ifdef COPY_PASTE
  int    w_slowpaste;		/* do careful writes to the window */
  int	 w_histheight;		/* all histbases are malloced with width * histheight */
  int	 w_histidx;		/* 0 <= histidx < histheight; where we insert lines */
  int	 w_scrollback_height;	/* number of lines of output stored, to be updated with w_histidx, w_histheight */
  struct mline *w_hlines;	/* history buffer */
  struct paster w_paster;	/* paste info */
#else
  int	 w_histheight;		/* always 0 */
#endif
  int	 w_pid;			/* process at the other end of ptyfd */
  int	 w_deadpid;		/* saved w_pid of a process that closed the ptyfd to us */

  char  *w_cmdargs[MAXARGS];	/* command line argument vector */
  char	*w_dir;			/* directory for chdir */
  char	*w_term;		/* TERM to be set instead of "screen" */

  int    w_lflag;		/* login flag */
  slot_t w_slot;		/* utmp slot */
#if defined (UTMPOK)
  struct utmp w_savut;		/* utmp entry of this window */
#endif

  char	 w_tty[MAXSTR];

  int    w_zauto;
#ifdef ZMODEM
  struct display *w_zdisplay;
#endif
#ifdef BUILTIN_TELNET
  struct sockaddr_storage w_telsa;
  char   w_telbuf[IOSIZE];
  int    w_telbufl;
  char   w_telmopts[256];
  char   w_telropts[256];
  int    w_telstate;
  char   w_telsubbuf[128];
  int    w_telsubidx;
  struct event w_telconnev;
#endif
  struct {
    int    on;    /* Is the alternate buffer currently being used? */
    struct mline *mlines;
    int    width;
    int    height;
#ifdef COPY_PASTE
    int    histheight;
    struct mline *hlines;
    int    histidx;
#else
    int histheight;	/* 0 */
#endif
    struct cursor cursor;
  } w_alt;

  struct event w_destroyev;	/* window destroy event */
#ifdef BSDWAIT
  union wait w_exitstatus;	/* window exit status */
#else
  int w_exitstatus;
#endif
};


#define w_encoding   w_layer.l_encoding
#define w_width  w_layer.l_width
#define w_height w_layer.l_height
#define w_x      w_layer.l_x
#define w_y      w_layer.l_y

/* definitions for w_type */
#define W_TYPE_PTY		0
#define W_TYPE_PLAIN		1
#define W_TYPE_TELNET		2
#define W_TYPE_GROUP		3


/*
 * Definitions for flow
 *   000  -(-)	flow off, auto would be off
 *   001  +(-)	flow on , auto would be off
 *   010  -(+)	flow off, auto would be on
 *   011  +(+)	flow on , auto would be on
 *   100  -	flow auto, currently off
 *   111  +	flow auto, currently on
 * Application controls auto_flow via TIOCPKT, if available, 
 * else via application keypad mode.
 */
#define FLOW_NOW	(1<<0)
#define FLOW_AUTO	(1<<1)
#define FLOW_AUTOFLAG	(1<<2)


/*
 * WIN gives us a reference to line y of the *whole* image
 * where line 0 is the oldest line in our history.
 * y must be in whole image coordinate system, not in display.
 */

#define WIN(y) ((y < fore->w_histheight) ? \
      &fore->w_hlines[(fore->w_histidx + y) % fore->w_histheight] \
    : &fore->w_mlines[y - fore->w_histheight])

#define Layer2Window(l) ((struct win *)(l)->l_bottom->l_data)

int WindowChangeNumber __P((int, int));
int OpenDevice __P((char **, int, int *, char **));

#endif /* SCREEN_WINDOW_H */

